src/lib/logging.ts

```diff
@@ -0,0 +1,453 @@
+     * @param logLevel
+     * @param message
+     */
+    public logWithId(
+        logLevel: string,
+        traceId: string,
+        message: string,
+        objects?: any,
+    ): void {
+        const level = this.levelToInt(logLevel);
+        if (level < this.minLevel) return;
+
+        const logEntry: LogEntry = {
+            level: logLevel,
+            module: this.module,
+            message,
+            objects,
+            traceId,
+        };
+
+        this.logManager.emit('log', logEntry);
+    }
+
+    public traceWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('trace', traceId, message, objects);
+    }
+    public debugWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('debug', traceId, message, objects);
+    }
+    public infoWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('info', traceId, message, objects);
+    }
+    public warnWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('warn', traceId, message, objects);
+    }
+    public errorWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('error', traceId, message, objects);
+    }
+}
+
+type TimingMap = {
+    // count, avg, min, max
+    [id: string]: number[];
+};
+
+const timingMap: TimingMap = {};
+
+/**
+ * This deceleration will log the time taken to run the function it is attached to. Be
+ * careful where it is added as it increases the output.
+ *
+ * @export
+ * @return {*}
+ */
+export const logCall = (
+    target: Object,
+    propertyKey: string,
+    descriptor: PropertyDescriptor,
+) => {
+    const originalMethod = descriptor.value;
+    //const logger = logging.getLogger('taskssql.perf');
+    descriptor.value = function (...args: any[]) {
+        const startTime = new Date(Date.now());
+        const result = originalMethod.apply(this, args);
+        const endTime = new Date(Date.now());
+        const name = `${target?.constructor?.name}${propertyKey}`;
+        const time = endTime.getTime() - startTime.getTime();
+        if (timingMap[name] === undefined) {
+            timingMap[name] = [];
+        }
+        timingMap[name].push(time);
+
+        //console.log(timingMap);
+        // if (endTime.getTime() - startTime.getTime() > 50) {
+        //     console.debug(
+        //         `[debug][taskssql.perf] ${String(timingMap[name].avg).padEnd(4)}${String(
+        //             endTime.getTime() - startTime.getTime(),
+        //         ).padEnd(4)} ${target?.constructor?.name.padEnd(10)}${propertyKey.padEnd(20)}`,
+        //     );
+
+        //     // logger.debug(
+        //     //     `${target?.constructor?.name}:${propertyKey}:called with ${args.length} arguments. Took: ${
+        //     //         endTime.getTime() - startTime.getTime()
+        //     //     }ms`,
+        //     // );
+        // }
+        return result;
```

Comment:
What needs to be done to enable this `taskssql.perf` code?

Reply:
It was pretty intensive in logging when enabled, left it in comment to come back when I have time to see if there is a better way to toggle it. Alternatively, it can be uncommented to get the information while debugging.